### ISSUE TITLE
Add a CS API to validate the user password

### ISSUE THREAD

**Issue by bmarty** [matrix-org/matrix-doc#2776]
**Is your feature request related to a problem? Please describe.**

As a client application, I want to ask the user to confirm its password before doing an action.

For instance, the user has forgotten its PIN code which protect access to the application or has entered several wrong PIN code.

In this case, the app could reset the PIN code, but after asking for the account password. For the moment, the application simply log out the user, which is a bit rought.

Current API only allow to validate the password by login to the account, but it has the side effect to create a new deviceId, which is not expected at all.

**Describe the solution you'd like**

Add a client server authenticated API
```
POST /_matrix/client/r0/password
```

With body

```json
{
  password="weakPassword"
}
```

and which will return 200 with empty Json in case of correct password, and 403 with a MatrixError in case of wrong password.

The API should be rate limited.

**Describe alternatives you've considered**

We should probably support all possible UIA flows, but not sure it is relevant for this use case, and not sure how we can implement that

**Additional context**

N/A

**Comment by clokep** [comment]
This essentially sounds like you want to kick off the user-interactive auth flow but then just receive whether it was successful or not at the end?

**Comment by bmarty** [comment]
Yes, it's quite a good point of view 👍

**Comment by Sorunome** [comment]
Yeah, using UIA would be better (and migrating normal login to UIA, too)

**Comment by uhoreg** [comment]
One way to do this without a specific endpoint is to perform a login flow and then immediately log out the access token.  (Except that it would be using login instead of UIA.)

**Comment by bmarty** [comment]
> One way to do this without a specific endpoint is to perform a login flow and then immediately log out the access token. (Except that it would be using login instead of UIA.)

We thought about that but it is really really hacky. It will create a device then remove it, this is not ideal. And what if the logout fail for any reason?
